src/common/btree.c

```diff
@@ -916,6 +917,62 @@ btr_root_grow(struct btr_context *tcx, TMMID(struct btr_node) mmid_left,
 	return 0;
 }
 
+struct btr_check_vbt {
+	TMMID(struct btr_node)	nd_mmid;
+	int			at;
+	uint32_t		intent;
+};
+
+static int
+btr_check_visibility(struct btr_context *tcx, struct btr_check_vbt *vbt)
+{
+	struct btr_record	*rec;
+	int			 rc;
+
+	if (btr_ops(tcx)->to_check_visibility == NULL)
+		return PROBE_RC_OK;
+
+	if (TMMID_IS_NULL(vbt->nd_mmid)) { /* compare the leaf trace */
+		struct btr_trace *trace = &tcx->tc_traces[BTR_TRACE_MAX - 1];
+
+		vbt->nd_mmid = trace->tr_node;
+		vbt->at = trace->tr_at;
+	}
+
+	if (!btr_node_is_leaf(tcx, vbt->nd_mmid))
+		return PROBE_RC_OK;
+
+	rec = btr_node_rec_at(tcx, vbt->nd_mmid, vbt->at);
+	rc = btr_ops(tcx)->to_check_visibility(&tcx->tc_tins, rec, vbt->intent);
+	if (rc == -DER_INPROGRESS) /* Unceration */
+		return PROBE_RC_INPROGRESS;
+
+	if (rc < 0) /* Failure */
+		return PROBE_RC_ERR;
+
+	switch (rc) {
+	case DTX_VBT_VISIBLE_DIRTY:
```

Comment:
Here is an example of the 'visibility' chaos. The record is actually 'invisible' from 2PC perspective, but it's 'visible' from caller's (aggregation code, which is iteration caller) perspective, and the term 'dirty' makes things even more confusing.

What about renaming to_check_visibility() to to_leaf_state(), and the return value can only be PREPARED, ABORTED or COMMITTED.